base/applications/kbswitch/kbsdll/kbsdll.c

```diff
@@ -1,94 +1,86 @@
 
-HHOOK hWinHook = NULL;
-HHOOK hShellHook = NULL;
-HINSTANCE hInstance = NULL;
-HWND hKbSwitchWnd = NULL;
+HINSTANCE g_hinstDLL = NULL;
+HHOOK g_hShellHook = NULL;
+HHOOK g_hCbtHook = NULL;
+HWND g_hwnd = NULL;
 
 static VOID
-PostMessageToMainWnd(UINT Msg, WPARAM wParam, LPARAM lParam)
+PostMessageToMainWnd(UINT uMsg, WPARAM wParam, LPARAM lParam)
 {
-    PostMessage(hKbSwitchWnd, Msg, wParam, lParam);
+    PostMessage(g_hwnd, uMsg, wParam, lParam);
 }
 
-LRESULT CALLBACK
-WinHookProc(int code, WPARAM wParam, LPARAM lParam)
+static LRESULT CALLBACK
+CbtProc(INT code, WPARAM wParam, LPARAM lParam)
 {
     if (code < 0)
-    {
-        return CallNextHookEx(hWinHook, code, wParam, lParam);
-    }
+        return CallNextHookEx(g_hCbtHook, code, wParam, lParam);
 
     switch (code)
     {
         case HCBT_SETFOCUS:
-        {
-            HWND hwndFocus = (HWND)wParam;
-            if (hwndFocus && hwndFocus != hKbSwitchWnd)
-            {
-                PostMessageToMainWnd(WM_WINDOW_ACTIVATE, wParam, lParam);
-            }
-        }
-        break;
+            PostMessageToMainWnd(WM_WINDOWSETFOCUS, wParam, lParam);
+            break;
     }
 
-    return CallNextHookEx(hWinHook, code, wParam, lParam);
+    return CallNextHookEx(g_hShellHook, code, wParam, lParam);
 }
 
-LRESULT CALLBACK
-ShellHookProc(int code, WPARAM wParam, LPARAM lParam)
+static LRESULT CALLBACK
+ShellProc(INT code, WPARAM wParam, LPARAM lParam)
 {
     if (code < 0)
-    {
-        return CallNextHookEx(hShellHook, code, wParam, lParam);
-    }
+        return CallNextHookEx(g_hShellHook, code, wParam, lParam);
 
     switch (code)
     {
         case HSHELL_LANGUAGE:
-        {
-            PostMessageToMainWnd(WM_LANG_CHANGED, wParam, lParam);
-        }
-        break;
+            PostMessageToMainWnd(WM_LANGUAGE, wParam, lParam);
+            break;
+        case HSHELL_WINDOWACTIVATED:
+            PostMessageToMainWnd(WM_WINDOWACTIVATED, wParam, lParam);
+            break;
+        case HSHELL_WINDOWCREATED:
+            PostMessageToMainWnd(WM_WINDOWCREATED, wParam, lParam);
+            break;
+        case HSHELL_WINDOWDESTROYED:
+            PostMessageToMainWnd(WM_WINDOWDESTROYED, wParam, lParam);
+            break;
     }
 
-    return CallNextHookEx(hShellHook, code, wParam, lParam);
+    return CallNextHookEx(g_hShellHook, code, wParam, lParam);
 }
 
-BOOL WINAPI
-KbSwitchSetHooks(VOID)
+BOOL APIENTRY KbsHook(HWND hwnd)
 {
-    hWinHook = SetWindowsHookEx(WH_CBT, WinHookProc, hInstance, 0);
-    hShellHook = SetWindowsHookEx(WH_SHELL, ShellHookProc, hInstance, 0);
+    g_hwnd = hwnd;
 
-    if (!hWinHook || !hShellHook)
+    g_hShellHook = SetWindowsHookEx(WH_SHELL, ShellProc, g_hinstDLL, 0);
+    g_hCbtHook = SetWindowsHookEx(WH_CBT, CbtProc, g_hinstDLL, 0);
+    if (!g_hShellHook || !g_hCbtHook)
     {
+        UnhookWindowsHookEx(g_hShellHook);
+        UnhookWindowsHookEx(g_hCbtHook);
+        g_hShellHook = g_hCbtHook = NULL;
```

Comment:
Should we keep `g_hwnd` initialized though? (If not, then I would suggest you move the `g_hwnd` assignation just before returning TRUE below.)

Reply:
`g_hwnd` is moved before returning `TRUE`.